constrain-lib/src/main/java/org/gradle/dependency/constrain/lib/serialize/ConstraintsJsonReader.java

```diff
@@ -0,0 +1,327 @@
+        // 2. Validate the JSON against the schema
+        validateAgainstJsonSchema(schema, json);
+        // 3. Convert the JSON to a JsonDependencyConstraints format.
+        final JsonDependencyConstraints constraints = readJsonDependencyConstraints(mapper, json);
+        // 4. Build the LoadedConstraints object, performing any additional validation.
+        final LoadedConstraints loadedConstraints = buildLoadedConstraints(constraints);
+        // 5. Generate a version of the input that is correctly formatted.
+        final List<String> formattedJson = generateFormattedJson(mapper, json);
+        // 6. Verify that the formatted JSON is the same as the input.
+        verifyNoFormattingDifferences(inputLines, formattedJson);
+        return loadedConstraints;
+    }
+
+    private static class InputStreamExtractedData {
+        private final JsonNode json;
+        private final List<String> inputLines;
+
+        private InputStreamExtractedData(JsonNode json, List<String> inputLines) {
+            this.json = json;
+            this.inputLines = inputLines;
+        }
+
+        public static InputStreamExtractedData create(ObjectMapper mapper, @WillClose InputStream in) {
+            final JsonNode json;
+            final List<String> inputLines;
+            try (BufferedReader useIn = new BufferedReader(new InputStreamReader(in))) {
+                useIn.mark(1 << 24); // Mark the stream so we can reset it later
+                // StreamReadFeature.AUTO_CLOSE_SOURCE was disabled when creating the ObjectFactory
+                json = mapper.readTree(useIn);
+                useIn.reset();
+                inputLines = useIn.lines().collect(Collectors.toList());
+                if (inputLines.isEmpty()) {
+                    throw new DependencyConstrainException("File is empty");
+                }
+            } catch (IOException ex) {
+                throw new DependencyConstrainException("Unable to read dependency constraints", ex);
+            }
+            return new InputStreamExtractedData(json, inputLines);
+        }
+    }
+
+    private static List<String> generateFormattedJson(ObjectMapper objectMapper, JsonNode json) {
+        final DefaultPrettyPrinter prettyPrinter = new DefaultPrettyPrinter();
+        monkeyPatchDefaultPrettyPrinter(prettyPrinter);
+        final DefaultIndenter indenter = new DefaultIndenter();
+        prettyPrinter.indentArraysWith(indenter);
+        prettyPrinter.indentObjectsWith(indenter);
+        final ObjectWriter writer = objectMapper.writer(prettyPrinter);
+        final String output;
+        try {
+            output = writer.writeValueAsString(json);
+        } catch (JsonProcessingException e) {
+            throw new DependencyConstrainException(
+                "Unable to generate formatted JSON for dependency constraints",
+                e
+            );
+        }
+        return Arrays.asList(output.split(System.lineSeparator()));
```

Comment:
you could move the return into the try, this would also get rid of needing to declare the uninitialized output variable outside.